curp/src/server/curp_node.rs

```diff
@@ -0,0 +1,759 @@
+use std::{collections::HashMap, fmt::Debug, sync::Arc, time::Duration};
+
+use clippy_utilities::NumericCast;
+use event_listener::Event;
+use futures::{pin_mut, stream::FuturesUnordered, StreamExt};
+use madsim::rand::{thread_rng, Rng};
+use parking_lot::{Mutex, RwLock, RwLockUpgradableReadGuard};
+use thiserror::Error;
+use tokio::{
+    sync::{broadcast, mpsc},
+    task::JoinHandle,
+    time::MissedTickBehavior,
+};
+use tracing::{debug, error, warn};
+use utils::config::ServerTimeout;
+
+use super::{
+    cmd_board::{CmdBoardRef, CommandBoard},
+    cmd_worker::start_cmd_workers,
+    gc::run_gc_tasks,
+    raw_curp::{AppendEntries, RawCurp, TickAction, Vote},
+    spec_pool::{SpecPoolRef, SpeculativePool},
+};
+use crate::{
+    cmd::{Command, CommandExecutor, ProposeId},
+    error::ProposeError,
+    message::ServerId,
+    rpc::{
+        self, connect::ConnectApi, AppendEntriesRequest, AppendEntriesResponse, FetchLeaderRequest,
+        FetchLeaderResponse, ProposeRequest, ProposeResponse, SyncError, VoteRequest, VoteResponse,
+        WaitSyncedRequest, WaitSyncedResponse,
+    },
+    TxFilter,
+};
+
+/// Uncommitted pool type
+pub(super) type UncommittedPool<C> = HashMap<ProposeId, Arc<C>>;
```

Comment:
I think your design is correct, but keeping two pools is more complicated than one. What do you think about it?

Reply:
I know it's possible to use spec_pool only with the leader inserting all cmds to it. But then the leader's spec_pool will violate the definition in the curp paper, i.e., all commands in spec_pool should be commutative. I am concerned about the correctness in recovery.